File: wrappers/js/examples/primitives.js
// Crypto and address primitives (offline).
//
// Run from wrappers/js:
//
//   LIB_DIR=../../core/build/native/nativeCompile
//   CCL_LIB_PATH=$LIB_DIR DYLD_LIBRARY_PATH=$LIB_DIR LD_LIBRARY_PATH=$LIB_DIR \
//     bun examples/primitives.js
import { CclBridge, TESTNET } from '../src/index.js';

const bridge = new CclBridge();
try {
  // --- Mnemonics ---
  const mnemonic = bridge.crypto.generateMnemonic(24);
  console.log('Generated 24-word mnemonic:', mnemonic);
  console.log('  valid?', bridge.crypto.validateMnemonic(mnemonic));
  console.log("  'not a real mnemonic' valid?", bridge.crypto.validateMnemonic('not a real mnemonic'));

  // --- Blake2b hashing (hex in -> hex out). "Hello" == 48656c6c6f ---
  console.log("Blake2b-256('Hello'):", bridge.crypto.blake2b256('48656c6c6f'));
  console.log("Blake2b-224('Hello'):", bridge.crypto.blake2b224('48656c6c6f'));

  // --- Ed25519 signing ---
  // getPrivateKey returns the 64-byte extended key; sign expects a 32-byte
  // Ed25519 key, so take the first 32 bytes (64 hex chars).
  const acct = bridge.account.create(TESTNET);
  const sk = bridge.account.getPrivateKey(acct.mnemonic, TESTNET).slice(0, 64);
  const pk = bridge.account.getPublicKey(acct.mnemonic, TESTNET);
  const messageHex = '68656c6c6f'; // "hello"
  console.log('Ed25519 signature:', bridge.crypto.sign(messageHex, sk));
  // A tampered signature is correctly rejected.
  console.log('  verify(fake signature) ->', bridge.crypto.verify('00'.repeat(64), messageHex, pk));

  // --- Address parsing & validation ---
  const addr = acct.base_address;
  console.log('Address valid?', bridge.address.validate(addr));
  console.log('Address info  :', bridge.address.info(addr));
  const raw = bridge.address.toBytes(addr);
  console.log('Address -> bytes -> address round-trips:', bridge.address.fromBytes(raw) === addr);
} finally {
  bridge.close();
}

File: wrappers/rust/examples/primitives.rs
//! Crypto and address primitives (offline).
//!
//! Run from wrappers/rust:
//!
//! ```text
//! LIB_DIR=../../core/build/native/nativeCompile
//! CCL_LIB_PATH=$LIB_DIR DYLD_LIBRARY_PATH=$LIB_DIR LD_LIBRARY_PATH=$LIB_DIR \
//!   cargo run --example primitives
//! ```
use ccl::{network, Bridge};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bridge = Bridge::new()?;

    // --- Mnemonics ---
    let mnemonic = bridge.crypto().generate_mnemonic(24)?;
    println!("Generated 24-word mnemonic: {}", mnemonic);
    println!("  valid? {}", bridge.crypto().validate_mnemonic(&mnemonic));
    println!(
        "  'not a real mnemonic' valid? {}",
        bridge.crypto().validate_mnemonic("not a real mnemonic")
    );

    // --- Blake2b hashing (hex in -> hex out). "Hello" == 48656c6c6f ---
    println!("Blake2b-256('Hello'): {}", bridge.crypto().blake2b_256("48656c6c6f")?);
    println!("Blake2b-224('Hello'): {}", bridge.crypto().blake2b_224("48656c6c6f")?);

    // --- Ed25519 signing ---
    // get_private_key returns the 64-byte extended key; sign expects a 32-byte
    // Ed25519 key, so take the first 32 bytes (64 hex chars).
    let created = bridge.account().create(network::TESTNET)?;
    let account: serde_json::Value = serde_json::from_str(&created)?;
    let mnemonic = account["mnemonic"].as_str().unwrap();
    let priv_ext = bridge.account().get_private_key(mnemonic, network::TESTNET, 0, 0)?;
    let pub_key = bridge.account().get_public_key(mnemonic, network::TESTNET, 0, 0)?;
    let message_hex = "68656c6c6f"; // "hello"
    let signature = bridge.crypto().sign(message_hex, &priv_ext[..64])?;
    println!("Ed25519 signature: {}", signature);
    // A tampered signature is correctly rejected.
    let fake_sig = "00".repeat(64);
    println!(
        "  verify(fake signature) -> {}",
        bridge.crypto().verify(&fake_sig, message_hex, &pub_key)
    );

    // --- Address parsing & validation ---
    let addr = account["base_address"].as_str().unwrap();
    println!("Address valid? {}", bridge.address().validate(addr));
    println!("Address info  : {}", bridge.address().info(addr)?);
    let raw = bridge.address().to_bytes(addr)?;
    println!(
        "Address -> bytes -> address round-trips: {}",
        bridge.address().from_bytes(&raw)? == addr
    );
    Ok(())
}
